exchange/bitswap/wantmanager.go

```diff
@@ -137,36 +137,49 @@ func (mq *msgQueue) runQueue(ctx context.Context) {
 	for {
 		select {
 		case <-mq.work: // there is work to be done
-
-			err := mq.network.ConnectTo(ctx, mq.p)
-			if err != nil {
-				log.Noticef("cant connect to peer %s: %s", mq.p, err)
-				// TODO: cant connect, what now?
-				continue
-			}
-
-			// grab outgoing message
-			mq.outlk.Lock()
-			wlm := mq.out
-			if wlm == nil || wlm.Empty() {
-				mq.outlk.Unlock()
-				continue
-			}
-			mq.out = nil
-			mq.outlk.Unlock()
-
-			// send wantlist updates
-			err = mq.network.SendMessage(ctx, mq.p, wlm)
-			if err != nil {
-				log.Noticef("bitswap send error: %s", err)
-				// TODO: what do we do if this fails?
-			}
+			mq.doWork(ctx)
 		case <-mq.done:
 			return
 		}
 	}
 }
 
+func (mq *msgQueue) doWork(ctx context.Context) {
+	// allow a minute for connections
+	// this includes looking them up in the dht
+	// dialing them, and handshaking
+	conctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
```

Comment:
On Thu, Jun 11, 2015 at 01:26:40PM -0700, Jeromy Johnson wrote:
> > +func (mq *msgQueue) doWork(ctx context.Context) {
> > +	// allow a minute for connections
> > +	// this includes looking them up in the dht
> > +	// dialing them, and handshaking
> > +	conctx, cancel := context.WithTimeout(ctx, time.Minute)
> > +	defer cancel()
> 
> alright, ten minutes then? tcp times out at 5.

Do we need explicit sub-part timeouts at all?  Can't the caller
specify their own timeout when they submit a task?  Looking into the
context handling here, runQueue gets launched with one context, and it
passes that context down to doWork.  But maybe the doWork context
should be a union of the runQueue context and a second context that
gets passed in via the channel?  With a check in runQueue to see if
any context-related doWork errors were due to a msgQueue-context
timeout/cancel (in which msgQueue should close down) or the
channel-passed context (in which case pass that error back to the
caller?).